fdbserver/Ratekeeper.actor.cpp

```diff
@@ -147,7 +148,11 @@ class RatekeeperImpl {
 		}
 	}
 
-	ACTOR static Future<Void> trackStorageServerQueueInfo(Ratekeeper* self, StorageServerInterface ssi) {
+	ACTOR static Future<Void> trackStorageServerQueueInfo(ResourceWeakRef<Ratekeeper> self,
+	                                                      StorageServerInterface ssi) {
+		if (!self.available()) {
```

Comment:
This feels a bit error prone, since we are required to check availability after every wait. It seems like there could be a better construct used within actors where we optimistically try to use the object and throw something like a cancellation error if the object is no longer available.

Reply:
I agree there could be a wrapper, so they throw something instead of manually testing every time. I am just feeling `operation_cancelled` is too broad and maybe another "surprise" in the code.

Reply:
I agree that we couldn't throw this error in the current generic class as is, because that could very easily be surprising. The way to resolve that would be to make the type more specific (e.g. something like the `WeakActorSelfPointer` proposal above) or to expose another accessor to the resource (e.g. `self.getOrCancel()->memberFunc()`). Because this particular use-case is common enough, I personally like the former as being a little easier to use. The latter could also work and still help to reduce risk because the availability check and the access would be combined into a single step, but it still would be possible to misuse the two step pattern.